### PR TITLE
UART failure with CONFIG_UART_ASYNC_API and DMA

### PR DESCRIPTION
**Pull request by FRASTM**
when enabling the DMA of the stm32f7 soc,  especially on the tests/drivers/uart/uart_async_api, 
the cache DMA is not coherent.
--> Same approach than the https://github.com/zephyrproject-rtos/zephyr/pull/27911/
This PR is to help for fixing of  the https://github.com/zephyrproject-rtos/zephyr/issues/31711
This still give an error on the `test_chained_write`  test case

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

File: drivers/serial/Kconfig.stm32 [hunk 1]
```diff
@@ -20,3 +20,17 @@ config UART_STM32
...
 	  Say y if you wish to use serial port on STM32 MCU.
+
+config UART_STM32_UNALIGNED_DMA
+	bool "STM32 MCU UART Unaligned DMA Support"
+	depends on UART_ASYNC_API
+	depends on CPU_CORTEX_M7
+	default y
+	help
+	  Support unaligned DMA buffer. When DMA buffers are not aligned on
+	  cache line boundaries, or there size is not a multiple of the
```
Comment (dcpleung) [inline drivers/serial/Kconfig.stm32:31]: s/there/their/

**Review by dcpleung — COMMENTED** [review]
There are couple commented out code block which needs to be removed.

File: drivers/serial/Kconfig.stm32 (hunk 1 above)
Comment (FRASTM) [reply drivers/serial/Kconfig.stm32:31]: @dcpleung I do not catch your point, sorry. This text is coming from the https://github.com/zephyrproject-rtos/zephyr/pull/27911/commits/70e111c1ae19e3aabfa1a95e906d40021542444b

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -668,6 +668,14 @@ static inline void async_evt_rx_rdy(struct uart_stm32_data *data)
 		uart_stm32_invalidate_cache(
 			data->dma_rx.buffer,
 			data->dma_rx.buffer_length);
+/*
+		if (data->dma_rx.buffer != event.data.rx.buf) {
+			memcpy(event.data.rx.buf,
+				data->dma_rx.dma_aligned_buffer,
+				data->dma_rx_buffer_len);
+		}
+*/
```
Comment (dcpleung) [inline drivers/serial/uart_stm32.c:677]: Please remove commented out code blocks.

File: drivers/serial/uart_stm32.c [hunk 3]
```diff
@@ -1017,13 +1026,29 @@ static int uart_stm32_async_tx(const struct device *dev,
...
+				buf_size = MIN(buf_size, addr_diff);
+			}
+
+			buf_size = MIN(buf_size, __SCB_DCACHE_LINE_SIZE);
+
+			memcpy(data->dma_tx.dma_aligned_buffer,
+					data->dma_tx.buffer, buf_size);
+			tx_data = data->dma_tx.dma_aligned_buffer;
+		}
+*/
```
Comment (dcpleung) [inline drivers/serial/uart_stm32.c:1051]: Remove commented out code block

File: drivers/serial/uart_stm32.c [hunk 4]
```diff
@@ -1077,6 +1102,28 @@ static int uart_stm32_async_rx_enable(const struct device *dev,
...
+			buf_size = UART_STM32_ROUND_DOWN(buf_size);
+		} else {
+			if (addr_diff != 0) {
+				buf_size = MIN(buf_size, addr_diff);
+			}
+
+			buf_size = MIN(buf_size, __SCB_DCACHE_LINE_SIZE);
+			rx_buf = data->dma_rx.dma_aligned_buffer;
+		}
+*/
```
Comment (dcpleung) [inline drivers/serial/uart_stm32.c:1124]: Remove commented out code block

**Review by dcpleung — CHANGES_REQUESTED** [review]

**Comment by FRASTM** [comment]
initial issue is now closed

**Pull request closed without merging**